addons/purchase/models/purchase.py

```diff
@@ -725,7 +725,16 @@ def _prepare_stock_moves(self, picking):
             'warehouse_id': self.order_id.picking_type_id.warehouse_id.id,
         }
         diff_quantity = self.product_qty - qty
-        if float_compare(diff_quantity, 0.0,  precision_rounding=self.product_uom.rounding) > 0:
+
+        # Take care that there will be no negative initial demand
+        moves = self.move_ids
+        pending_moves = moves.filtered(lambda r: r.state not in ['done', 'cancel'])
+        done_moves = moves.filtered(lambda r: r.state == 'done') - pending_moves
```

Comment:
Removing of the pending moves is not needed as it is explicitly not possible to be part of...same for the other code on the sale order part.